add POV and other v1.8, v1.7 changes

nvkelso opened the pull request:
- ⭐️Add POV for boundaries and capitals; add map unit lines (eg Wales and Scotland)
    - ![tilezen_v1d8_disputed_boundaries](https://raw.githubusercontent.com/tilezen/vector-datasource/master/docs/images/tilezen-v1.8.0-disputed-boundaries.gif)
- 🚚Add heavy goods vehicle (hgv) overlay
- 🤑Add toll road overlay
- 🚧 Add construction roads
- 🏞 Add new landuse treatments
- 👨 Fix `aboriginal_lands` boundary line treatment
- 🔋 Fix `generator` icons to use kind_detail lookup
- 🐯 Fix zoo `enclosure` labels
- 🚘 Fix motorway junction (exit) labels
- ☠️ Fix `danger_area` and `range` landuse styling (not RED!)
- 🏷 Update `labels-11` theme to not reference all the client side hacks that were fixed server side
- 💻 Remove bunch of logic that moved server side (like for filtering out bad features, setting label collision priority)
- 🎉 Upgrade Tangram JS to v0.19

nvkelso committed:
add POV and other v1.8, v1.7 changes

nvkelso committed:
Merge branch 'gh-pages' into nvkelso/v1d8-changes

nvkelso committed:
restore missing building style from bad web merge

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -12,8 +12,96 @@ global:
+                    return false;
+            }
+        }
+    ux_point_of_view_kind_filter_country_capital: |
```
@bcamper Can you take a look at this section? When viewing Taipei the capital icon should turn OFF when in the Chinese POV as `country_capital:cn` is **false** (even though the default `country_capital` is **true**). The same pattern we used for the name fallbacks doesn't work here.

bcamper replied:
Believe this is fixed in e5e6f31075230c78fff1d2acf53bd064fa8811b8. The issue is that we don't want the *absence* of a `capital_country:XX` value to turn a capital *off*. We want to check for `!= null` so that we only consider values that are explicitly `true` or `false`.

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -1032,34 +1141,18 @@ layers:
             filter:
                 all:
                     - kind: [lake, water, riverbank, reservoir, swimming_pool]
-                any:
```
This is all taken care of server side now.

bcamper replied:
🤘🤖

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -1175,39 +1273,18 @@ layers:
                         - $geometry: point
                         - not: { kind: [riverbank, dock, playa] }
                         - area: true
-                    any:
```
all water filtering is server side now

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -1118,18 +1211,23 @@ layers:
                     lines:
                         color: [[8,[0.667,0.792,0.824]],[11,[0.698,0.808,0.835]],[13,[0.710,0.796,0.824]],[14,[0.522,0.792,0.855]]]
                         width: [[9,0.5px],[10,0.75px],[11,0.75px],[12,0.8px],[13,1px],[14,1px],[15,1px],[16,1px],[17,1.5px]]
+#            intermittent: 
```
Not sure this should be removed... but it wasn't active here...

nvkelso replied:
Restored in 84584c5b55e8c4126ab45ac39c8d15ecc821380b.

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -1175,39 +1273,18 @@ layers:
                                 family: global.text_font_family
                                 style: italic
                                 weight: normal
-                                size: 12px
```
This is fixed server side now.

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -1263,20 +1340,6 @@ layers:
                         - $geometry: point
                         - kind: playa
                         - area: true
-                    any:
```
Moved server side.

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -1382,11 +1446,96 @@ layers:
                 text_source: global.ux_language_text_source
                 visible: false    # labels are enabled by each layer below
+                priority: |
+                    function() { return (feature['min_zoom'] + (1 - 1 / feature['collision_rank']) * 0.1) || 65; }
```
Use the v1.7 `collision_rank` for predictable POI labeling. 

Really we should also have a server budget and do some collisions there for low- and mid-zooms.

bcamper replied:
yesss

Cool to see this. But yes also makes sense to incur more cost server-side in the future where possible.

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -1382,11 +1446,96 @@ layers:
+#                   order: 1000
+#                   width: 1px
+#                   visible: true
+#        z-hgv:
```
This should be turned into a new overlay in the DAT gui.

nvkelso replied:
Fixed in be9020163d755c795db104b0fb05bf5d4144d44d.

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -1493,7 +1642,7 @@ layers:
                 draw:
                     text-blend-order:
-                        priority: 50
+#                        priority: 50
```
These should just be removed... they are all set above now from the data.

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -3154,13 +3317,16 @@ layers:
-                    - admin_level: 2
+                    #- kind: [country, indefinite, indeterminate]
+                    #- admin_level: 2
+                    - global.ux_point_of_view_kind_filter_country
```
Added new POV filter for v1.7 and v1.8

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -3176,7 +3342,9 @@ layers:
                 draw:
                     text-blend-order:
-                        priority: 2
+#                        priority: 2
```
Remove this line, rely on the next line for server priority

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -3194,27 +3362,79 @@ layers:
-            draw:
+            filter: 
+                global.ux_point_of_view_kind_filter_disputed_etc
+                #- kind: [disputed, lease_limit, line_of_control, overlay_limit]
```
Can remove these 2 lines, they are carried in the global now

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -3194,27 +3362,79 @@ layers:
+                #- kind: [disputed, lease_limit, line_of_control, overlay_limit]
+                #- kind: [disputed_breakaway, disputed_claim, disputed_elusive, disputed_reference_line]
+            draw:
+                #lines: 
```
This lines style were just for debug, remove them

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -3194,27 +3362,79 @@ layers:
-                filter: { kind: [disputed, line_of_control] }
-                draw:
+                filter: 
+                    #- kind: [disputed, line_of_control]
```
Remove these lines, they are carried in the global now.

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -3194,27 +3362,79 @@ layers:
+                    #- kind: [disputed_breakaway, disputed_claim, disputed_elusive, disputed_reference_line]
+                    global.ux_point_of_view_kind_filter_disputed_star
+                draw:
+                    #lines: 
```
Remove this lines debug style

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -3194,27 +3362,79 @@ layers:
                         #width: [[1, 0.2px], [2, 0.5px], [4, 1.0px], [9, 1.5px], [14, 2.5px], [16, 3.5px], [17, 10m]]
                         width: [[1, 0.5px], [2, 1.0px], [4, 1.5px], [9, 2.5px], [14, 3.5px], [16, 4.5px], [17, 14m]]
+                        #width: 20px
+#            disputed_star_early:
```
This seems like it's junk and should be removed?

nvkelso replied:
Fixed in 0c2624e3cb61fceb0555934313ffbccd251a0be5.

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -3194,27 +3362,79 @@ layers:
+#                    dashed_big:
+#                        width: [[5, 0px], [5, 1.0px], [9, 2.5px], [14, 3.5px], [16, 4.5px], [17, 14m]]
+
+        unrecognized:
```
New for POV boundaries

bcamper replied:
Interesting -- out of curiosity, what's a test case for this?

nvkelso replied:
Unrecognized country boundaries in Morocco, Israel, Kashmir areas.

Unrecognized region boundaries in Kashmir area.

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -3194,27 +3362,79 @@ layers:
+                 dashed_big:
+                     visible: false
+
+        map_unit:
```
Added a new kind (like for England, Wales)

bcamper replied:
huh, cool

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -3232,11 +3452,21 @@ layers:
                     lines:
                         visible: false
+
+            aboriginal_lands:
```
EG around 10/41.1154/-483.5195

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -3285,7 +3538,7 @@ layers:
             draw:
                 text-blend-order:
-                    priority: 3
+#                    priority: 3
```
These old priority should be removed.

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -3920,7 +4225,7 @@ layers:
                     # then selectively reveal in classes below (ultimately a data problem)
                     visible: global.text_visible_neighbourhoods_e
-                    priority: 29
+#                    priority: 29
```
Remove old priority

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -4042,291 +4357,75 @@ layers:
-                draw:
-                    mapzen_icon_library:
-                        visible: false
+#        no-name:
```
This section should be removed, it's all server side now and working.

nvkelso replied:
Fixed via 8cbbc7f15e702d7df72ecfcfd0df8e32d170e4fa.

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -4042,291 +4357,75 @@ layers:
-            draw:
-                mapzen_icon_library:
-                    priority: 41
+#        has-area:
```
Should also be removed.

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -4457,19 +4558,19 @@ layers:
-            draw:
-                mapzen_icon_library:
-                    priority: 43
+#        landuse-labels-not-any-above:
```
Remove

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -4652,45 +4758,45 @@ layers:
-#                           }
-#                       }
-#                   }
+#                #debug for polygon merging
```
This building section needs revisiting... I had a bad merge conflict along the way.

nvkelso replied:
Seems fine, actually.

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -4959,13 +5075,13 @@ layers:
-                draw:
-                    dots:
-                        order: 237
+            #shipyard:
```
Remove this since fixed on server

nvkelso replied:
Fixed in 0fb72303a983c8c27d45fb2c7be9a941e0138dc3.

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -5071,6 +5185,56 @@ layers:
                 draw:
                     dots:
                         color: [0.584,0.843,0.792]
+        grass:
```
New landcover colors for new data

bcamper committed:
update capital POV logic

nvkelso committed:
fix motorway_junction labels; disable landuse catchall (doesnt work)

nvkelso committed:
move to prod tiles now they are released

nvkelso committed:
remove unreferenced style for transparent dots

nvkelso committed:
restore intermitent water styling

nvkelso commented on bubble-wrap-style.yaml:
```diff
@@ -655,6 +744,16 @@ styles:
                     color.rgb = mix(color.rgb, vec3(0.850), TileDots(45., 0.1));
                     // size was set to 35 previously, now 45
 
+    dots_transparent:
```
Removed this unused section in 942074aacc99e58a82c8dfc080f4c4125a8093fc.

nvkelso committed:
remove outdated priority, set all to data

nvkelso committed:
cleanup old boundaries code

nvkelso committed:
remove z_country_capital_pov debug

nvkelso committed:
remove logic that moved to server

nvkelso committed:
remove shipyard (now serverside)

nvkelso committed:
minor

nvkelso committed:
simplify sources

nvkelso committed:
move toll and hgv to overlays

nvkelso committed:
fix hgv tunnels

nvkelso commented:
@bcamper This PR is cleaned up and ready for new 👀, please!

bcamper commented on bubble-wrap-style.yaml:
```diff
@@ -1175,39 +1259,18 @@ layers:
-                        - { $zoom: { min: 15 }, area: { min: 50000 } }
-                        - { $zoom: { min: 16 }, area: { min: 20000 } }
-                        - { $zoom: { min: 17 } }
+                        - not: { kind: [sea, ocean] }
```
Can't this be merged into the `not: { kind: [...] }` filter above now?

nvkelso replied:
The parent filter still needs to allow those 2 kinds... and then the sibling layer (right below this one) targets them for a custom style.

bcamper commented on bubble-wrap-style.yaml:
```diff
@@ -1382,11 +1431,111 @@ layers:
+            filter:
+                all:
+                    - hgv: true
+                    - function() { return global.sdk_hgv_overlay; }
```
On the Tangram side, we could add more direct support for `globals` in layer filters in the future, which would avoid the function compile/call. I actually did some work on this but never merged it, could consider for v0.20:

https://github.com/tangrams/tangram/compare/filter-globals
https://github.com/tangrams/tangram/tree/filter-boolean

Those would support patterns like:

```
filter:
  all:
    - hgv: true
    - global.sdk_hgv_overlay # this is a straight value substitution evaluating to true/false
```

```
filter:
  hgv: true
  $global.sdk_hgv_overlay # special $global syntax needed here, otherwise `global` will be interpreted as a feature property
```

nvkelso replied:
Supporting globals in layer tests directly instead of via function would be stellar :)

bcamper commented on bubble-wrap-style.yaml:
```diff
@@ -1382,11 +1431,111 @@ layers:
+                    - function() { return global.sdk_hgv_overlay; }
+            draw:
+                lines:
+                    visible: true
```
Superfluous (defaults to true)

nvkelso replied:
Fixed in 68bc27fb5a5056d7ed7b63d3b7660477fc4c135a.

bcamper commented on bubble-wrap-style.yaml:
```diff
@@ -1382,11 +1431,111 @@ layers:
+                    visible: true
+                    color: blue
+                    width: 1.7px
+            z-none:
```
`z-none` and `z-restricted` filters are the same, is one supposed to be a different value?

bcamper replied:
Still wondering about this one.


NVK: Removed in 930a049bc25e74953cfdf0e9bf4ca5d04dd8ea94.

bcamper commented on bubble-wrap-style.yaml:
```diff
@@ -1382,11 +1431,111 @@ layers:
+                    draw:
+                        lines:
+                            # color: [[16, [0.843,0.843,0.843]]]
+                            color: purple
```
As written, this draw block doesn't do anything since it just re-sets the color to `purple` (which should already be `purple` from parent...?

nvkelso replied:
This was needed because it "folds" in with the general roads layer styling which has a more specific selector for road tunnels which reset the purple coloring in this layer's parent. Fun times.

bcamper commented on bubble-wrap-style.yaml:
```diff
@@ -1118,18 +1197,23 @@ layers:
-                        color: [[11,[0.671,0.788,0.812]],[13,[0.710,0.800,0.824]],[14,[0.635,0.812,0.843]]]
-                        width: [[9,0px],[11,0px],[12,0px],[13,0.5px],[14,0.75px],[15,0.75px],[16,0.75px],[17,1px]]
+                        color: [[12,[0.671,0.788,0.812]],[14,[0.710,0.800,0.824]],[15,[0.635,0.812,0.843]]]
+                        width: [[9,0px],[11,0px],[13,0px],[14,0.5px],[15,0.75px],[16,0.75px],[17,1px]]
```
Can drop the extra `0px` stops now (anything <= z13 will be 0px):

`width: [[13,0px],[14,0.5px],[15,0.75px],[16,0.75px],[17,1px]]`

bcamper commented on bubble-wrap-style.yaml:
```diff
@@ -1382,11 +1431,111 @@ layers:
+                                        return ('generic_shield-' + feature.hgv_restriction_shield_text.length + 'char');
+                                    }
+                                }
+                            size: [[7,[75%,75%]],[12,[100%,100%]],[15,[110%,110%]]]
```
Should only need one % since it's the same for width/height:

`size: [[7, 75%], [12, 100%], [15, 110%]]`

bcamper commented on bubble-wrap-style.yaml:
```diff
@@ -1382,11 +1431,111 @@ layers:
+                            priority: 1
+                            # Size is specified below, and limited to 5 char in another filter
+                            repeat_group: shields
+                            repeat_distance: [[8,90px],[10,40px],[13,30px]]
```
I had to check in the code if `repeat_distance` supports zoom stops.. it does! Supports functions too. :) But the docs don't reflect either (updating).

@matteblair @tallytalwar does ES support zoom stops and/or functions for `repeat_distance`? Let's file an issue for parity if not.

nvkelso replied:
See: https://github.com/tangrams/tangram-es/issues/2079.

bcamper commented on bubble-wrap-style.yaml:
```diff
@@ -1382,11 +1431,111 @@ layers:
+                                - [12, 1]
+                                - [13, 1.50]
+                                - [14, 2.0]
+                            #cull_from_tile: true
```
This is a deprecated JS keyword, we can remove entirely.

bcamper commented on bubble-wrap-style.yaml:
```diff
@@ -1382,11 +1431,111 @@ layers:
+                            repeat_group: shields
+                            repeat_distance: [[8,90px],[10,40px],[13,30px]]
+                            placement: midpoint
+                            color: purple
```
May be superfluous depending on parent layer/filter questions above.

bcamper commented on bubble-wrap-style.yaml:
```diff
@@ -1382,11 +1431,111 @@ layers:
+                                - [13, 1.50]
+                                - [14, 2.0]
+                            #cull_from_tile: true
+                            visible: true
```
Superfluous as parent layers currently written.

bcamper commented on bubble-wrap-style.yaml:
```diff
@@ -1493,19 +1642,15 @@ layers:
                     $zoom: [13,14]
                 draw:
                     text-blend-order:
-                        priority: 50
```
✂️ ❤️

bcamper commented on bubble-wrap-style.yaml:
```diff
@@ -2265,7 +2390,6 @@ layers:
                 draw:
                     mapzen_icon_library:
-                        priority: 47
                         #color: yellow
```
Not part of the diff but.. not needed :)

bcamper commented on bubble-wrap-style.yaml:
```diff
@@ -2389,9 +2506,8 @@ layers:
                     }
                 draw:
-                    mapzen_icon_library:
+                    mapzen_icon_library: {}
```
What's the deal with these lines (here and below) setting empty an empty object for the `draw` block? Can these just be removed or is that doing something? I'd expect it would just be a no-op when the layers are merged together.

nvkelso replied:
I set to `draw: mapzen_icon_library: {}`  because it's not obvious to me if I can have a sublayer of a parent layer without a draw block (next few lines down). So it was just me being caution & clever at the same time.

bcamper commented on bubble-wrap-style.yaml:
```diff
@@ -2407,9 +2523,8 @@ layers:
                     }
                 draw:
-                    mapzen_icon_library:
+                    mapzen_icon_library: {}
```
See comment above on similar.

bcamper commented on bubble-wrap-style.yaml:
```diff
@@ -2490,6 +2603,13 @@ layers:
+            construction:
+                filter: { kind: construction }
+                draw:
+                    dashed:
```
We *can* move the `dash` inline into the `draw` groups now, but probably outside (already large) scope of this PR. But for future, if we want to do that instead of having a separate style.

nvkelso replied:
Agree... though Tangram ES not supporting that kinda puts a damper on that / map design is broken without the dashing.

bcamper replied:
Oh right, forgot it's not in ES yet.

bcamper commented on bubble-wrap-style.yaml:
```diff
@@ -3154,13 +3275,14 @@ layers:
                 order: global.feature_order
-#                color: red
-#                width: [[9, 1px], [14, 2px], [16, 3px], [17, 8m]]
+                #debug:
```
Is keeping this useful? (Fine if so.)

nvkelso replied:
Yes... I keep some debugging in when I keep going back to debug (see buildings and roads).

bcamper commented on bubble-wrap-style.yaml:
```diff
@@ -3578,328 +3745,302 @@ layers:
                             size: [[10,8px],[11,0px]]
                             sprite: capital-l
-                            priority: 5
-                        text-blend-order:
```
Was the `text-blend-order` removed because these are handled by the icon drawing?

nvkelso replied:
It's because the master layer's default `text-blend-order` sets `priority: global.collision_priority` above, so they aren't needed every instance – since all the specific values moved server side and it's just a lookup now.

bcamper commented on bubble-wrap-style.yaml:
```diff
@@ -4021,6 +4161,14 @@ layers:
+            filter: { kind: generator }
+            draw:
+                mapzen_icon_library:
+                    sprite: function() { return 'generator_' & feature.kind_detail; }
```
I think you want `'generator_' + feature.kind_detail` here. You don't want `'generator_' & feature.kind_detail` (single `&` is a bitwise numeric operator) or `'generator_' && feature.kind_detail` (which would always return just `generator_`).

bcamper commented on bubble-wrap-style.yaml:
```diff
@@ -4041,293 +4189,6 @@ layers:
                         visible: false
         #end different icon sprites
 
-        #most POI icons without name should be hidden, except these
```
Ah is this tied to the `text-blend-order` removals above? (Or maybe I've lost track.)

nvkelso replied:
This was all moved server side, and the diff here makes that hard to read.

bcamper commented on bubble-wrap-style.yaml:
```diff
@@ -4414,13 +4274,14 @@ layers:
-                        font:
-                            style: italic
-                            weight: normal
+                text:
```
Just to confirm, this `text` is supposed to be an alternative to a `mapzen_icon_library` icon, which is turned off? Elsewhere we use that designated `text-blend-order` style for such cases, right?

nvkelso replied:
Well, funny you should ask. If I move to `text-blend-order` instead of `text` around 50% of the labels are culled when I expect them to be shown :\

http://localhost:8000/#18/37.73212/-122.50216

![image](https://user-images.githubusercontent.com/853051/62810569-27229180-bab3-11e9-80dc-89f5cb22ac13.png)

nvkelso replied:
Well, I don't know why but this other suggestion of yours is the root cause of the missing text:

>-                    size: [[7,[75%,75%]],[12,[100%,100%]],[15,[110%,110%]]]
>+                    size: [[7,[75%]],[12,[100%]],[15,[110%]]]

bcamper commented on bubble-wrap-style.yaml:
```diff
@@ -4498,6 +4350,11 @@ layers:
                     mapzen_icon_library:
                         visible: false
+            low-priority-early-z14-icons:
+                filter: { kind_tile_rank: { min: 50 }, $zoom: [13,14] }
```
Interesting, haven't seen that `kind_tile_rank` in action before.

bcamper commented on bubble-wrap-style.yaml:
```diff
@@ -4802,6 +4664,7 @@ layers:
                 draw:
                     dots:
                         color: [[4,[0.525,0.765,0.686]],[9,[0.416,0.737,0.631]]]
+        # TODO: This seems broken
```
Tile content, or something I can help with?

nvkelso replied:
That's a comment from a while ago, not sure what's going on there without some debugging that's beyond this PR.

bcamper commented on bubble-wrap-style.yaml:
```diff
@@ -5120,6 +5036,19 @@ layers:
                     color: [0.765,0.765,0.765]
                     visible: true
 
+#        # (20190729: nvk) TODO: This should only show when no other kind condition
```
I believe this kind of "else" condition can be done with the new `exclusive` layer property added in JS -- make all the other layers `exclusive`, and if none matches, the non-exclusive "else" one does -- but it's not in ES yet.

nvkelso replied:
YES! `exclusive` would be great here :) So in Tangram ES the layer won't match and won't display? Seems backwards compatible and okay to implement?

nvkelso replied:
Reading the docs in https://github.com/tangrams/tangram/pull/705 I'm confused how I'd implement this. Is there a `!exclusive` or `priority: -1` to mean evaluate this one layer last? Or do I have to add priority to every other sibling layer here, too?

bcamper commented on index.html:
```diff
@@ -22,12 +22,13 @@
+        <!-- <script src="https://mapzen.com/tangram/0.19/tangram.debug.js"></script> -->
         <!-- dev -->
+        <!-- <script src="https://line-label-repeats--tangram.netlify.com/dist/tangram.debug.js"></script> -->
         <!-- <script src="https://precog.mapzen.com/tangrams/tangram/master/dist/tangram.debug.js"></script> -->
```
going wayyy back ;)

bcamper commented on index.html:
```diff
@@ -393,6 +394,55 @@
+                'United States': 'us',
+                'Vietnam': 'vn'
+            };
+            // use query language, else default to English
```
comment isn't relevant, from old copy-paste

nvkelso replied:
Fixed via d7defa5ef35a5bd1dae1e9d3a79da5b602715337.

bcamper reviewed (commented):
Great new features and I love all the code removal!

There are a few questions around filter logic or values that are worth addressing, may be one or two unintentional items, but otherwise all looks good.

I'd like to pull out a couple of these oft-repeated lines into globals, to be more concise and lower risk of copy-paste errors, such as:

`var kind = (global.ux_point_of_view && feature['kind:'+global.ux_point_of_view]) || (global.ux_point_of_view_fallback && feature['kind:'+global.ux_point_of_view_fallback]) || feature.kind;`

and

```
priority: |
                    function() { return (feature['min_zoom'] + (1 - 1 / feature['collision_rank']) * 0.1) || 65; }
```

I can take a crack at that.

bcamper committed:
pull POV feature kind logic into separate global function

bcamper committed:
pull collision rank logic into separate global function

bcamper commented:
OK I extracted some repetitive logic in:

f591907cbf414cab6beb4538295b53b92ff1705c
91f3b8f294d41bb124942474bf935b6a002b9a6b

Noticing that the pre-existing language toggle stuff could benefit from this as well, but it looks more complicated because of the many variants (short name, left/right borders, etc.), so I'll leave that for now.

bcamper commented:
Actually now that I'm looking at these lines like:

```
priority: |
  function() { return (feature['min_zoom'] + (1 - 1 / feature['collision_rank']) * 0.1) || 65; }
```

I'm realizing the default value of 65 will never be triggered. If `collision_rank` is missing, you'll get `-Infinity`. If `min_zoom` is missing, you'll still get a number, but less than `1`.

What's the desired fallback logic here?

bcamper commented:
So is this the desired collision priority logic:

- If both `min_zoom` and `collision_rank` exist, use the formula.
- If only `min_zoom` exists, use it (or use `min_zoom + 1` if features without a `collision_rank` should rank lower by default than those that do).
- If neither exist, use the provided default.

But, even then, the fallback values in the style like 65, 58, etc. don't make sense as final priority values when compared with `min_zoom` based values that are going to be in the ~1-20 range.

Are those numbers supposed to be defaults for `collision_rank`?

nvkelso committed:
fix comment about POV

nvkelso committed:
remove superfilous visible: true

nvkelso commented:
You're right... those fallback values are a holdover from before all the `collision_rank` were moved server side and spread out with much greater detail. But something like this is still desired. If a feature has a `min_zoom` but no `collision_rank` then it should be evaluated as _worst_ among peers of the same `min_zoom`.

Example:

- Feature A:
    - `min_zoom`: 10
    - `collision_rank`: 1000
- Feature B:
    - `min_zoom`: 10
    - `collision_rank`: 1001
- Feature C:
    - `min_zoom`: 10
    - `collision_rank`: null

The outcome should be Feature **A** wins over **B** wins over **C**.

I think this could be accomplished by falling back to `feature.min_zoom + 0.999` and call it good?

nvkelso committed:
rework new collision_priority function; add custom for roads and transit

nvkelso committed:
pesky zoom 14 labels

nvkelso committed:
add collision for water

nvkelso commented:
I've addressed the collision priority changes primarily in 92afcabb466831752f87ca73d17668dd5b57df8f (simplified to remove the defaults), and added logic for **road**, **water**, and **transit** labels (turn transit overlay on to see it in action).

nvkelso commented:
@bcamper This is ready for another review, please. I've addressed the collision priority stuff you called out (and fixed the transit overlay labels in the process), but don't know how to proceed on the https://github.com/tangrams/bubble-wrap/pull/277#discussion_r311863310 so maybe we defer that to a later PR?

bcamper commented on bubble-wrap-style.yaml:
```diff
@@ -5392,28 +5411,125 @@ layers:
+                interactive: false
+                color: [[13,[0,0,0.5,0.25]],[15,[0,0,0.5,0.8]]]
+                width: [[11,0.5px],[12,0.8px],[16,1.5px],[18,1.5m]]
+                # let roads sort themselves past zoom 14
```
This is just assigning an explicit order value though, is the comment incorrect? (or the code?)

nvkelso replied:
Updated the comment in 1da2c657047b2b9d6ba99306a5fa0ee6010f466e.

We moved the zoom changes server side a few releases ago so the comment was outdated. There look to be a few static int values in the scene file still. I've filed https://github.com/tangrams/bubble-wrap/issues/278 to track that.

bcamper commented on bubble-wrap-style.yaml:
```diff
@@ -5392,28 +5411,125 @@ layers:
         filter:
             kind: [station, tram_stop, bus_station, subway_entrance, halt, stop, platform, bus_stop, stop_position ]
             any:
                 - area: false
```
What's going on with both `area: false` and `area: true`?

nvkelso replied:
That's odd, probably a holdout from around v1.0 tiles? 

In any event, removing it doesn't break anything and simplifies the style: eafad2608ad9a4ba8510daabd71556d99c504928.

bcamper committed:
collision priority needs to check for null

(otherwise it'll divide by null rather than falling back to the desired value)

bcamper reviewed (approved):
Made an additional fix to the those collision priority functions -- they need to check for `null` first, otherwise they'll divide by `null` which will be `Infinity` instead of falling back to the desired value.

Couple other small questions about things that may have issues, but probably not significant/blocking (please do take a look though).

bcamper commented:
> @bcamper This is ready for another review, please. I've addressed the collision priority stuff you called out (and fixed the transit overlay labels in the process), but don't know how to proceed on the [#277 (comment)](https://github.com/tangrams/bubble-wrap/pull/277#discussion_r311863310) so maybe we defer that to a later PR?

Agree we can follow up on this later.

bcamper commented:
> You're right... those fallback values are a holdover from before all the `collision_rank` were moved server side and spread out with much greater detail. But something like this is still desired. If a feature has a `min_zoom` but no `collision_rank` then it should be evaluated as _worst_ among peers of the same `min_zoom`.
> ...
> I think this could be accomplished by falling back to `feature.min_zoom + 0.999` and call it good?

Yep that works, made some updates in https://github.com/tangrams/bubble-wrap/pull/277/commits/8fdfe3be79e87e5bb48e355a7f30c19c5ba14a3d.

nvkelso committed:
update comment per PR review

nvkelso committed:
remove janky layer

nvkelso committed:
fix collision_priority (was reversed); address PR comment; emphasize largest transit stations

nvkelso reviewed (commented):
I fixed a noticeable regression in the style between prod and this PR by changing order of operations in the `collision_priority` function in eafad2608ad9a4ba8510daabd71556d99c504928, which was preventing neighbourhood labels in SF to show up at zooms less than 16 and was preventing Golden Gate Park label from showing up w/r/t roads labels, and too many "bay" labels north of San Francisco.

nvkelso committed:
few other PR comments from Brett
